server/src/main/java/org/elasticsearch/transport/InboundPipeline.java

```diff
@@ -0,0 +1,175 @@
+import org.elasticsearch.common.util.PageCacheRecycler;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.function.BiConsumer;
+
+public class InboundPipeline implements Releasable {
+
+    private static final InboundMessage PING_MESSAGE = new InboundMessage(null, true);
+
+    private final InboundDecoder decoder;
+    private final InboundAggregator aggregator;
+    private final BiConsumer<TcpChannel, InboundMessage> messageHandler;
+    private final BiConsumer<TcpChannel, Tuple<Header, Exception>> errorHandler;
+    private ArrayDeque<ReleasableBytesReference> pending = new ArrayDeque<>(2);
+    private boolean isClosed = false;
+
+    public InboundPipeline(Version version, PageCacheRecycler recycler, BiConsumer<TcpChannel, InboundMessage> messageHandler,
+                           BiConsumer<TcpChannel, Tuple<Header, Exception>> errorHandler) {
+        this(new InboundDecoder(version, recycler), new InboundAggregator(), messageHandler, errorHandler);
+    }
+
+    private InboundPipeline(InboundDecoder decoder, InboundAggregator aggregator,
+                            BiConsumer<TcpChannel, InboundMessage> messageHandler,
+                            BiConsumer<TcpChannel, Tuple<Header, Exception>> errorHandler) {
+        this.decoder = decoder;
+        this.aggregator = aggregator;
+        this.messageHandler = messageHandler;
+        this.errorHandler = errorHandler;
+    }
+
+    @Override
+    public void close() {
+        isClosed = true;
+        Releasables.closeWhileHandlingException(decoder, aggregator);
+        Releasables.closeWhileHandlingException(pending);
+        pending.clear();
+    }
+
+    public void handleBytes(TcpChannel channel, ReleasableBytesReference reference) throws IOException {
+        pending.add(reference.retain());
+
+        final ReleasableBytesReference composite;
+        if (pending.size() == 1) {
+            composite = pending.peekFirst();
+        } else {
+            final ReleasableBytesReference[] bytesReferences = pending.toArray(new ReleasableBytesReference[0]);
+            final Releasable releasable = () -> Releasables.closeWhileHandlingException(bytesReferences);
+            composite = new ReleasableBytesReference(new CompositeBytesReference(bytesReferences), releasable);
+        }
+
+        int bytesConsumed = 0;
+        final ArrayList<Object> fragments = new ArrayList<>();
+        boolean continueHandling = true;
+
+        while (continueHandling && isClosed == false) {
+            boolean continueDecoding = true;
+            while (continueDecoding) {
+                final int remaining = composite.length() - bytesConsumed;
+                if (remaining != 0) {
+                    try (ReleasableBytesReference slice = composite.retainedSlice(bytesConsumed, remaining)) {
+                        final int bytesDecoded = decoder.decode(slice, fragments::add);
```

Comment:
Why are we adding to a list here and not instantly passing things on? Does this not mean that we need to decode (and possibly decompress) the full slice before reacting to the request (e.g. invoke a breaker)?

Reply:
We're attempting to decode the entire message in a single go since that will be the most common scenario. 

The exact path for how messages can be circuit broken will be built out in follow-up work. Compression is going to be kind of an issue no matter what since we have to decompress to know the decompressed size (or put some type of header indicating the final size).